network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/ConnectDisconnectUtil.java

```diff
@@ -0,0 +1,182 @@
+     * @param isTypeSwitchToOperate type of switches that can be operated
+     * @param reportNode report node
+     * @return {@code true} if all the specified terminals have been connected, else {@code false}.
+     */
+    static boolean connectAllTerminals(AbstractIdentifiableImpl<?, ?> identifiable, List<Terminal> terminals, Predicate<Switch> isTypeSwitchToOperate, ReportNode reportNode) {
+
+        // Booleans
+        boolean isAlreadyConnected = true;
+        boolean canBeConnected = true;
+
+        // Initialisation of a list to open in case some terminals are in node-breaker view
+        Set<SwitchImpl> switchForConnection = new HashSet<>();
+
+        // We try to connect each terminal
+        for (Terminal terminal : terminals) {
+            // Check if the terminal is already connected
+            if (terminal.isConnected()) {
+                // If the terminal is already connected, log and continue on other terminals
+                reportNode.newReportNode()
+                    .withMessageTemplate("alreadyConnectedTerminal", "A terminal of identifiable ${identifiable} is already connected.")
+                    .withUntypedValue("identifiable", identifiable.getId())
+                    .withSeverity(TypedValue.WARN_SEVERITY)
+                    .add();
+                continue;
+            } else {
+                isAlreadyConnected = false;
+            }
+
+            // If it's a node-breaker terminal, the switches to connect are added to a set
+            if (terminal instanceof TerminalImpl<?> terminalImpl && terminalImpl.isNodeBeakerTopologyKind()) {
+                canBeConnected = terminalImpl.getConnectingSwitches(isTypeSwitchToOperate, switchForConnection);
+            }
+            // If it's a bus-breaker terminal, there is nothing to do now
+
+            // Exit if the terminal cannot be connected
+            if (!canBeConnected) {
+                return false;
+            }
+        }
+
+        // Exit if the connectable is already fully connected
+        if (isAlreadyConnected) {
+            return false;
+        }
+
+        // Connect all terminals
+        for (Terminal terminal : terminals) {
+            if (!terminal.isConnected()
+                && terminal instanceof TerminalImpl<?> terminalImpl
+                && !terminalImpl.isNodeBeakerTopologyKind()) {
+                terminalImpl.connectBusBreaker();
+            }
+        }
+
+        // The switches are now closed in node-breaker terminals
+        closeSwitches(identifiable.getIndex(), switchForConnection);
+
+        return true;
+    }
+
+    public static void closeSwitches(NetworkObjectIndex index, Set<SwitchImpl> switchesToClose) {
+        Set<String> closedSwitches = new HashSet<>();
+        switchesToClose.forEach(switchImpl -> {
+            switchImpl.setOpen(false);
+            index.updateSwitchResource(switchImpl.getResource());
```

Comment:
switchImpl.setOpen already updates the resource and notify the update, no ?

Reply:
No idea, I kept this part as it was before (I just regrouped it in the forEach loop as @antoinebhs asked).
See [this commit](https://github.com/powsybl/powsybl-network-store/commit/0332613965e02ad2c056558d4659f9f7ce61c873#diff-2c8d37c896cd472a0bb985851b2afa7e4cab7bdb7212028d06cf153e9f74ac85) if you want to see how it was last year